pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go

```diff
@@ -0,0 +1,218 @@
+package delegate
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metainternal "k8s.io/apimachinery/pkg/apis/meta/internalversion"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/klog/v2"
+
+	"github.com/openshift/api/oauth"
+	oauthv1 "github.com/openshift/api/oauth/v1"
+
+	oauthapi "github.com/openshift/oauth-apiserver/pkg/oauth/apis/oauth"
+	accesstokenregistry "github.com/openshift/oauth-apiserver/pkg/oauth/apiserver/registry/oauthaccesstoken/etcd"
+	oauthprinters "github.com/openshift/oauth-apiserver/pkg/oauth/printers/internalversion"
+	"github.com/openshift/oauth-apiserver/pkg/printers"
+	"github.com/openshift/oauth-apiserver/pkg/printerstorage"
+)
+
+// rest implements a RESTStorage for access tokens against etcd
+type REST struct {
+	accessTokenStorage *accesstokenregistry.REST
+
+	tableConvertor rest.TableConvertor
+}
+
+// we only allow retrieving the tokens => no Create() or Update()
+var _ rest.Lister = &REST{}
+var _ rest.Getter = &REST{}
+var _ rest.Watcher = &REST{}
+var _ rest.GracefulDeleter = &REST{}
+var _ rest.Scoper = &REST{}
+
+// NewREST returns a RESTStorage object that will work against access tokens
+func NewREST(accessTokenStorage *accesstokenregistry.REST) (*REST, error) {
+	return &REST{
+		accessTokenStorage: accessTokenStorage,
+		tableConvertor:     printerstorage.TableConvertor{TableGenerator: printers.NewTableGenerator().With(oauthprinters.AddOAuthOpenShiftHandler)},
+	}, nil
+}
+
+func (r *REST) New() runtime.Object {
+	return &oauthv1.UserOAuthAccessToken{}
+}
+
+func (r *REST) NewList() runtime.Object {
+	return &oauthv1.UserOAuthAccessTokenList{}
+}
+
+func (r *REST) List(ctx context.Context, options *metainternal.ListOptions) (runtime.Object, error) {
+	ctxUserName, err := getUserFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	sanitizedListOpts, err := userFieldSelectorToListOptions(options, ctxUserName)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenListGenericObj, err := r.accessTokenStorage.List(ctx, sanitizedListOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenList, ok := tokenListGenericObj.(*oauthapi.OAuthAccessTokenList)
+	if !ok {
+		return nil, errors.NewInternalError(fmt.Errorf("failed to convert generic accesstoken LIST result to its typed version"))
```

Comment:
could we be more specific and log the underlying type for `tokenListGenericObj` and `OAuthAccessTokenList` ?

Reply:
This error gets propagated to the user, does it not? I think it might be confusing when they see a mention of a different type from what they requested

Reply:
When the conversion can fail? Can it only happen when the wiring is wrong? 

In that case, I would prefer to see a very clear and specific error so that I can debug it.